usb: add support for dedicated work queue

pabigot opened the pull request:
Add an optional work queue to mitigate deadlocks when the USB infrastructure (servicing CDC-ACM UART) is blocked by non-USB tasks on the system work queue that invoke UART operations assuming these can progress without putting the task to sleep.

Note per https://github.com/zephyrproject-rtos/zephyr/issues/27599#issuecomment-692214683 that fixes isolated to the CDC-ACM driver are not sufficient, as usb-transfer must also be able to make progress.

Fixes #27599

pabigot committed:
usb: fix header language linkage

The block that introduced a C linkage region was not closed in the
header as it should be.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
usb: clean CMakeLists conditions

Use the helper function to conditionally include source files.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
usb: add optional work queue

The USB infrastructure currently uses the system work queue for
offloading transfers, CDC-ACM UART transmission/reception, and device
firmware activities.  This causes problems when the system work queue
is also used to initiate some activities (such as UART) that normally
complete without requiring an external thread: in that case the USB
infrastructure is prevented from making progress because the system
work queue is blocked waiting for the USB infrastructure to provide
data.

Break the dependency by allowing the USB infrastructure to use a
dedicated work queue which doesn't depend on availability of the
system work queue.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests: bluetooth: shell: add usb shell test configuration

Verify builds and execution when a CDC-ACM UART is used for the shell.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

Vudentz commented on tests/bluetooth/shell/src/main.c:
```diff
@@ -121,6 +122,11 @@ static void hrs_notify(void)
 void main(void)
 {
+	if (IS_ENABLED(CONFIG_USB_UART_CONSOLE)) {
+		usb_enable(NULL);
```
Couldn't USB auto enable when console is used? That should enable logging the init sequence as well.

pabigot replied:
No.  The user may want to register a custom handler.  (I do in real apps.)

pabigot committed:
usb: cdc_acm: fix formatting specifiers in log messages

The z length modifier must be applied to avoid build errors when the
format is verified by the compiler on systems where size_t is larger
than int.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

emob-nordic commented on subsys/usb/Kconfig:
```diff
@@ -114,6 +114,38 @@ config USB_MAX_POWER
 
+config USB_WORKQUEUE
+	bool "Use a dedicate work queue in the USB subsystem"
+	default y if USB_CDC_ACM
```
Shouldn't we limit this a bit? I think there is no good reason to force dedicated work queue for all CDC_ACM samples and increase RAM usage if could be avoided. The issues are visible when CDC_ACM is used as a backed for other subsystems which also use system work q for scheduling their work. I think this, at least for the time being, should be enabled only if `CONFIG_MCUMGR_ON_DEV_NAME="CDC_ACM_n"`, where n - number of CDC_ACM device. The problem is (afaik) we are unable to compare strings within Kconifgs. The idea of how to address such problem I described in [comment](https://github.com/zephyrproject-rtos/zephyr/pull/28339#issuecomment-691927108) for #28339. The question is if we are okay with accepting increasing the resources needed or not, if yes, then skip this comment.

jfischer-phytec-iot replied:
> Shouldn't we limit this a bit?

Well it is not "forced", you can still disable it. I think it should be `default y if USB_CDC_ACM`

pabigot replied:
I don't want things that work with "real" (threadless) UARTs to break when switched to CDC-ACM.  IMO this is the safest way to do that.  We already know shell with bluetooth, and mcumgr, independently run into this problem.

We can take this out, but then when people expand a working sample to add code that starts to use the system work queue it will stop working, and much time will be wasted figuring out why.

emob-nordic commented on subsys/usb/usb_work_q.c:
```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2018 Intel Corporation
+ * Copyright (c) 2016 Wind River Systems, Inc.
```
Should those be there if this is a new file? Just asking.

pabigot replied:
It's basically a copy of `kernel/system_work_q.c` with a few identifiers changed, so I preserved those copyrights.

eobalski reviewed (approved):
LGTM, just 2 comments because I usually pick on.
